lib/src/validator/relative_version_numbering.dart

```diff
@@ -0,0 +1,84 @@
+
+import '../entrypoint.dart';
+import '../exceptions.dart';
+import '../package_name.dart';
+import '../pubspec.dart';
+import '../validator.dart';
+
+/// Gives a warning when publishing a new version, if the latest published
+/// version lower to this was not opted into null-safety.
+class RelativeVersionNumberingValidator extends Validator {
+  static const String guideUrl = 'https://dart.dev/null-safety/migration-guide';
+  static const String semverUrl =
+      'https://dart.dev/tools/pub/versioning#semantic-version';
+
+  final String _server;
+
+  RelativeVersionNumberingValidator(Entrypoint entrypoint, this._server)
+      : super(entrypoint);
+
+  @override
+  Future<void> validate() async {
+    final hostedSource = entrypoint.cache.sources.hosted;
+    List<PackageId> existingVersions;
+    try {
+      existingVersions = await hostedSource
+          .bind(entrypoint.cache)
+          .getVersions(hostedSource.refFor(entrypoint.root.name, url: _server));
+    } on PackageNotFoundException {
+      existingVersions = [];
+    }
+    existingVersions..sort((a, b) => a.version.compareTo(b.version));
+    final previousVersion = existingVersions.lastWhere(
+        (id) =>
+            !id.version.isPreRelease && id.version < entrypoint.root.version,
+        orElse: () => null);
+    if (previousVersion == null) return;
+
+    final previousPubspec =
+        await hostedSource.bind(entrypoint.cache).describe(previousVersion);
+
+    final currentOptedIn = _optedIntoNullSafety(entrypoint.root.pubspec);
+    final previousOptedIn = _optedIntoNullSafety(previousPubspec);
+
+    if (currentOptedIn && !previousOptedIn) {
+      hints.add('You are about to publish a package opting into null-safety.\n'
+          'The latest version ${previousVersion.version} has not opted in.\n'
+          'Be sure to read $guideUrl for best practices.');
+    } else if (!currentOptedIn && previousOptedIn) {
+      hints.add(
+          'You\'re about to publish a package that doesn\'t opt into null safety,\n'
+          'but the previous version (${previousVersion.version}) was opted in.\n'
+          'This change is likely to be backwards incompatible. See $semverUrl');
+    }
+  }
+
+  static bool _optedIntoNullSafety(Pubspec pubspec) {
+    final sdkConstraint = pubspec.originalDartSdkConstraint;
+
+    /// If the sdk constraint is not a `VersionRange` something is wrong, and
+    /// we cannot deduce the language version.
+    ///
+    /// This will hopefully be detected elsewhere.
+    ///
+    /// A single `Version` is also a `VersionRange`.
+    if (sdkConstraint is! VersionRange) return false;
+    final constraintMin = (sdkConstraint as VersionRange).min;
+
+    if (constraintMin == null) return false;
+
+    final languageVersion =
+        Version(constraintMin.major, constraintMin.minor, 0);
+
+    return languageVersion >= _firstVersionSupportingNullSafety;
+  }
+
+  static final _firstVersionSupportingNullSafety = Version.parse('2.10.0');
```

Comment:
Maybe we should make a `LanguageVersion` class that we can use to represent this... because this is a bit scary.

If we update this to something else than `x.y.0` (adding a prerelease or build version number) then we introduce a bug.
It's probably best that we don't store LanguageVersion numbers in `Version` objects.

I'm okay with refactoring this later because we're already doing this elsewhere.